Updates to VEP 140 and GA NAD hotplug feature

frenzyfriday opened the pull request:
### VEP Metadata

**Tracking issue**: https://github.com/kubevirt/enhancements/issues/140
**SIG label**: /SIG network

### What this PR does
- We dont want to restrict the NAD swap just to bridge bindings, but mention
that only scenarios with bridge binding will be tested.
- Explain why Alpha phase is skipped
- Amend the section about annotation patching
- Choosing Option 1 as the preferred implementation
- Proposes GA for the feature in v1.9

### Special notes for your reviewer

Addresses: https://github.com/kubevirt/enhancements/pull/138#pullrequestreview-3601612221

frenzyfriday committed:
Updates to VEP 140

- We dont want to restrict the NAD swap just to bridge bindings, but
mention that only scenarios with bridge binding will be tested.
The implementation is binding agnostic.
- Explain why Alpha phase is skipped
- Amend the section about annotation patching
- Choosing Option 1 as the preferred implementation

Signed-off-by: Ananya Banerjee <anbanerj@redhat.com>

orelmisan reviewed (approved):
Thank you!

EdDev commented on veps/sig-network/hotpluggable-nad-ref.md:
```diff
@@ -184,9 +186,12 @@ The following scenarios are to be added to the e2e tests:
 ## Graduation Requirements
 
+### Alpha
+- Alpha stage will be skipped as most of the required logic already exists.
```
I think you need to be a bit more convincing.
Perhaps you can be more specific on why it is safe to do so. E.g. no API change is introduced, depends on 1, 2, 3 functionalities that already server a, b, c hotplug changes... etc.

frenzyfriday commented on veps/sig-network/hotpluggable-nad-ref.md:
```diff
@@ -75,9 +75,11 @@ multus annotation of the pod based on the difference between the VMI interface s
 in order to not break the existing hot-plug / hot-unplug flows.
 
+This is the preferred option as this lies withing the scope of the project and can be changed later on to support In Place NAD swap if the behaviour in DNC changes.
+
```
Ack, should I change it to `and can be changed  to support In Place NAD swap if in the future the behaviour in DNC changes` ?

nirdothan replied:
Very good! 
However small nit:
> if the relevant issue in DNC is closed in the future.

Consider "resolved" or "fixed" rather than "closed".

nirdothan reviewed (approved):
Thank you!

EdDev commented on veps/sig-network/hotpluggable-nad-ref.md:
```diff
@@ -181,12 +183,18 @@ This enhancement relies on the existing KubeVirt capabilities and does not intro
+### Alpha
+This enhancement proposes starting at Beta stage, skipping Alpha, based on the following:
+- No API changes: It would modify low level logics concerning conditions for restart required and migration, copying selected items from VM to VMI and conditions for patching pod multus annotation. 
+- Low risk: It leverages live migration which already exists and would follow a similar workflow as hotplug/unplug or network interfaces, which has been stable in production.
```
nit:
> hotplug/unplug or network

Is the "or" misplaced?

EdDev reviewed (approved):
Thanks!

xpivarc commented on veps/sig-network/hotpluggable-nad-ref.md:
```diff
@@ -181,12 +183,18 @@ This enhancement relies on the existing KubeVirt capabilities and does not intro
+### Alpha
+This enhancement proposes starting at Beta stage, skipping Alpha, based on the following:
+- No API changes: It would modify low level logics concerning conditions for restart required and migration, copying selected items from VM to VMI and conditions for patching pod multus annotation. 
+- Low risk: It leverages live migration which already exists and would follow a similar workflow as hotplug/unplug of network interfaces, which has been stable in production.
```
Not true, actually this feature is like nuclear weapon, a lot can go wrong with it and only advanced users should use it.

frenzyfriday replied:
I updated the wording to say `- Low risk: It leverages live migration which already exists and would follow a similar workflow as hotplug/unplug of network interfaces, which has been stable in production. Changing networks, which was allowed previously through a restart, will now be possible without restart.`
Is that better?

EdDev commented on veps/sig-network/hotpluggable-nad-ref.md:
```diff
@@ -181,12 +183,18 @@ This enhancement relies on the existing KubeVirt capabilities and does not intro
 
+### Alpha
+This enhancement proposes starting at Beta stage, skipping Alpha, based on the following:
+- No API changes: It would modify low level logics concerning conditions for restart required and migration, copying selected items from VM to VMI and conditions for patching pod multus annotation. 
```
The semantics is the VEP, but the point here is that there is no need to experiment on the API, i.e. there is nothing to examine in terms of extensibility, clarity, etc.
In this regard, the sentence is not trying to say anything about the semantics.

Said all that, I understand your point. "API change" includes semantics, therefore, clarification that it is just about API fields on the spec would have been nicer.

frenzyfriday replied:
I updated the wording to say `- No changes in API fields`. Is that better?

frenzyfriday committed:
VEP 140: GA NAD live update

This commit proposes to GA the NAD hotplug feature. Periodic tests have
been running and green for more than 2 weeks.

Signed-off-by: Ananya Banerjee <anbanerj@redhat.com>

frenzyfriday commented:
/remove-lifecycle stale

frenzyfriday commented:
Last [change](https://github.com/kubevirt/enhancements/compare/1fae9df8f3a7ce5fac338bc8a15bde9a4210f530..d1935d77484ef09436d2146ae507367ae96ef007) is a rebase + adds a bit more clarification to the `### Alpha` section

nirdothan commented on veps/sig-network/hotpluggable-nad-ref.md:
```diff
@@ -84,9 +84,11 @@ multus annotation of the pod based on the difference between the VMI interface s
 without removing `nad-with-vlan10` causing a conflict in mac and interface id and thus failing the attachment.
 
-To prevent that, the reconciliation logic in virt-controller will be adjusted to not patch the annotation when only the NAD reference of an existing interface is updated, 
+To prevent that, the reconciliation logic in virt-controller will be adjusted to not patch the annotation when existing parameters of existing interface are edited, 
```
Why do we need this change? and what does "edited" mean for a controller?

frenzyfriday replied:
I meant that the network annotation will not be patched if any interface parameter (not only the NAD ref) is updated by the user (edited). It will be done only for hotplug/unplugs

frenzyfriday replied:
Removed this change [here](https://github.com/kubevirt/enhancements/compare/e1258747afefbcc6522df47396e21f67522d5273..1559c9d12b31ee4ff358bd87584e3b271b5ecc18) 
we can revisit it later

frenzyfriday commented:
Last change addresses Nir's [comment](https://github.com/kubevirt/enhancements/pull/166#discussion_r3146909458)

orelmisan commented on veps/sig-network/hotpluggable-nad-ref.md:
```suggestion
- [X] (R) GA target version is explicitly mentioned and approved
```

frenzyfriday replied:
Ack, updated [here](https://github.com/kubevirt/enhancements/compare/c8626854342c8ef80dc10dddf629127645fdaa7d..312324e803e11611bb544bca2e883a9668e929dc)

orelmisan commented on veps/sig-network/hotpluggable-nad-ref.md:
```diff
@@ -190,12 +192,18 @@ This enhancement relies on the existing KubeVirt capabilities and does not intro
 
 The following scenarios are to be added to the e2e tests:
 - Change the NAD reference on a VM and verify that the VM is now connected to the new network and has connectivity.
+- E2E testing will be done only with bridge binding. Additional bindings would be considered after GA.
```
Let's drop this sentence please. We will expand the e2e test suite according to our needs.

frenzyfriday replied:
Ack, updated [here](https://github.com/kubevirt/enhancements/compare/df7a938dca7e5e65d3a7857c78d8c8a2714f8c55..c8626854342c8ef80dc10dddf629127645fdaa7d)

orelmisan commented on veps/sig-network/hotpluggable-nad-ref.md:
Please add a note that integration with the Dynamic Networks Controller will be deferred to when the need arises.

frenzyfriday replied:
Ack, added [here](https://github.com/kubevirt/enhancements/compare/1559c9d12b31ee4ff358bd87584e3b271b5ecc18..df7a938dca7e5e65d3a7857c78d8c8a2714f8c55)

orelmisan reviewed (commented):
Thanks for the PR @frenzyfriday!                                                                                                                            Broadening the scope beyond bridge binding makes sense to me as the implementation is binding-agnostic, so it should work for all bindings relevant for secondary networks.

frenzyfriday commented:
Last changes addresses Orel's [comments](https://github.com/kubevirt/enhancements/pull/166#pullrequestreview-4182161284)

orelmisan reviewed (approved):
Thank you @frenzyfriday!
I support graduating this feature to GA based on the experience gathered so far.

EdDev reviewed (approved):
Thank you!

I’m aware of this feature being exercised on our functional e2e test suite successfully and on d/s setups by users.

I believe we are safe to proceed to GA.

vladikr commented:
/approve
Ok, I'm fine with GAing as well.
